File: ml/utils/losses.py
import numpy as np 
def mse(true,prediction):
    return np.mean((np.array(true)-np.array(prediction))**2)
def mse_derivative(prediction,true):
    return (2*(prediction-true)/len(true))
def binary_cross_entropy(prediction,true):
    return -1*(np.mean(true*np.log(prediction)+(1-true)*np.log(1-prediction)))
def bce_derivative(prediction, true):
    return (prediction - true) / (prediction * (1 - prediction) * len(true))

File: ml/ml_cpp/utils/losses.cpp
#include "losses.hpp"
double bce(const Eigen :: MatrixXd& predction, const Eigen ::MatrixXd& true_value);
double mse(const Eigen::MatrixXd& predictions , const Eigen::MatrixXd&true_val);
Eigen::MatrixXd mse_derivative(const Eigen::MatrixXd& prediction, const Eigen::MatrixXd& true_value);

double mse(const Eigen::MatrixXd& predictions , const Eigen::MatrixXd&true_val){
    return (predictions-true_val).array().square().mean();
}
Eigen::MatrixXd mse_derivative(const Eigen::MatrixXd& prediction, const Eigen::MatrixXd& true_value){
    return 2*(prediction-true_value)/true_value.rows();
}

double bce(const Eigen :: MatrixXd& predction, const Eigen ::MatrixXd& true_value){
    return -1*(true_value.array()*(Eigen::log(predction.array())) + (1-true_value.array())*(Eigen::log(1-predction.array()))).mean();
}